### DISCUSSION TITLE
Casadi Symbolic Framework result show

### DISCUSSION THREAD

**Discussion by huihuishen** [casadi/casadi#3132] [Q&A]
My program got a symbolic result. I use python's print, and there are a lot of @ symbols in the result display. I can understand that this is done to make the result more efficient, but I need a complete symbolic result now. Don't use The @ symbol defines some reusable variables, how should I set the output to get the symbol result formula without the @ symbol?

**Comment by huihuishen** [comment]
> 
![image](https://user-images.githubusercontent.com/55009515/233399831-6a4b037c-487b-4401-bd6e-0c1db02a5bab.png)

**Comment by jaeandersson** [comment]
CasADi symbolic expressions allow shared subexpressions. That is, the same subexpression can be used multiple times in the same expression. If you write something like:
```python
x = SX.sym('x')
z = x
for i in range(100): z = sin(z) * z
print(z)
```
then printing it without "@" would require some 2^100 entries. With "@", you only need in the order 200 nodes. So it reflects better how the expression is actually stored in CasADi and ensures that the printout is compact, also when there are shared subexpressions. The expressions with "@" in the printout are precisely the ones which appears more than once, either in the matrix itself or in the intermediate expressions used to define the matrix.

**Reply by huihuishen** [reply]
Thank for your explanation. But I just want get  the expression without shared subexpressions for some use, so what can I do to get the answer?

**Reply by huihuishen** [reply]
> CasADi symbolic expressions allow shared subexpressions. That is, the same subexpression can be used multiple times in the same expression. If you write something like:
> 
> ```python
> x = SX.sym('x')
> z = x
> for i in range(100): z = sin(z) * z
> print(z)
> ```
> 
> then printing it without "@" would require some 2^100 entries. With "@", you only need in the order 200 nodes. So it reflects better how the expression is actually stored in CasADi and ensures that the printout is compact, also when there are shared subexpressions. The expressions with "@" in the printout are precisely the ones which appears more than once, either in the matrix itself or in the intermediate expressions used to define the matrix.


https://github.com/casadi/casadi/issues/1408#issuecomment-96600655
The less  compact printout is what i need. But I dot know how to do. Please help. Thank you!!!

**Reply by huihuishen** [reply]
> CasADi symbolic expressions allow shared subexpressions. That is, the same subexpression can be used multiple times in the same expression. If you write something like:
> 
> ```python
> x = SX.sym('x')
> z = x
> for i in range(100): z = sin(z) * z
> print(z)
> ```
> 
> then printing it without "@" would require some 2^100 entries. With "@", you only need in the order 200 nodes. So it reflects better how the expression is actually stored in CasADi and ensures that the printout is compact, also when there are shared subexpressions. The expressions with "@" in the printout are precisely the ones which appears more than once, either in the matrix itself or in the intermediate expressions used to define the matrix.

Or store the full expression to a file

**Reply by jaeandersson** [reply]
> > CasADi symbolic expressions allow shared subexpressions. That is, the same subexpression can be used multiple times in the same expression. If you write something like:
> > ```python
> > x = SX.sym('x')
> > z = x
> > for i in range(100): z = sin(z) * z
> > print(z)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > then printing it without "@" would require some 2^100 entries. With "@", you only need in the order 200 nodes. So it reflects better how the expression is actually stored in CasADi and ensures that the printout is compact, also when there are shared subexpressions. The expressions with "@" in the printout are precisely the ones which appears more than once, either in the matrix itself or in the intermediate expressions used to define the matrix.
> 
> [#1408 (comment)](https://github.com/casadi/casadi/issues/1408#issuecomment-96600655) The less compact printout is what i need. But I dot know how to do. Please help. Thank you!!!

We don't support i such printout anymore, as explained in that issue. You can write your own string processing code if you want.

**Reply by jaeandersson** [reply]
> > CasADi symbolic expressions allow shared subexpressions. That is, the same subexpression can be used multiple times in the same expression. If you write something like:
> > ```python
> > x = SX.sym('x')
> > z = x
> > for i in range(100): z = sin(z) * z
> > print(z)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > then printing it without "@" would require some 2^100 entries. With "@", you only need in the order 200 nodes. So it reflects better how the expression is actually stored in CasADi and ensures that the printout is compact, also when there are shared subexpressions. The expressions with "@" in the printout are precisely the ones which appears more than once, either in the matrix itself or in the intermediate expressions used to define the matrix.
> 
> Or store the full expression to a file

The compact printout *is* the full expression. CasADi expressions are graphs not trees.

**Reply by huihuishen** [reply]
> > > CasADi symbolic expressions allow shared subexpressions. That is, the same subexpression can be used multiple times in the same expression. If you write something like:
> > > ```python
> > > x = SX.sym('x')
> > > z = x
> > > for i in range(100): z = sin(z) * z
> > > print(z)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > then printing it without "@" would require some 2^100 entries. With "@", you only need in the order 200 nodes. So it reflects better how the expression is actually stored in CasADi and ensures that the printout is compact, also when there are shared subexpressions. The expressions with "@" in the printout are precisely the ones which appears more than once, either in the matrix itself or in the intermediate expressions used to define the matrix.
> > 
> > 
> > [#1408 (comment)](https://github.com/casadi/casadi/issues/1408#issuecomment-96600655) The less compact printout is what i need. But I dot know how to do. Please help. Thank you!!!
> 
> We don't support i such printout anymore, as explained in that issue. You can write your own string processing code if you want.

Can you give some idea program?
